src/opnsense/mvc/app/controllers/OPNsense/Base/ApiMutableTableModelControllerBase.php

```diff
@@ -0,0 +1,183 @@
+        $result = array("result"=>"failed");
+        if ($this->request->isPost() && $this->request->hasPost(static::$internalModelName))) {
+            $mdl = $this->getModel();
+            // FIXME Is this correct?
+            $node = $this->getNodes()->add();
+            $node->setNodes($this->request->getPost(static::$internalModelName)));
+            // FIXME What do we do with this? Is this traffic-shaper specific? Do we need a hook here?
+            $node->origin = "TrafficShaper"; // set origin to this component.
+            return $this->save($mdl, $node, static::$internalModelName));
+        }
+        return $result;
+    }
+
+    /**
+     * delete item by uuid
+     * @param $uuid item unique id
+     * @return array status
+     */
+    public function delItemAction($uuid)
+    {
+        $result = array("result"=>"failed");
+        if ($this->request->isPost()) {
+            $mdl = $this->getModel();
+            if ($uuid != null) {
+                if (getNodes()->del($uuid)) {
+                    // if item is removed, serialize to config and save
+                    $mdl->serializeToConfig();
+                    Config::getInstance()->save();
+                    $result['result'] = 'deleted';
+                } else {
+                    $result['result'] = 'not found';
+                }
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * toggle item by uuid (enable/disable)
+     * @param $uuid item unique id
+     * @param $enabled desired state enabled(1)/disabled(1), leave empty for toggle
+     * @return array status
+     */
+    public function toggleItemAction($uuid, $enabled = null)
+    {
+        $result = array("result" => "failed");
+        if ($this->request->isPost()) {
+            $mdl = $this->getModel();
+            if ($uuid != null) {
+                $node = $mdl->getNodeByUUID($uuid);
+                if ($node != null) {
+                    if ($enabled == "0" || $enabled == "1") {
+                        $node->enabled = (string)$enabled;
+                    } elseif ($node->enabled->__toString() == "1") {
+                        $node->enabled = "0";
+                    } else {
+                        $node->enabled = "1";
+                    }
+                    $result['result'] = $node->enabled;
+                    // if item has toggled, serialize to config and save
+                    $mdl->serializeToConfig();
+                    Config::getInstance()->save();
+                }
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * search items
+     * @return array
+     */
+    public function searchItemsAction()
+    {
+        $this->sessionClose();
+        $mdl = $this->getModel();
+        $grid = new UIModelGrid($this->getNodes());
+        return $grid->fetchBindRequest(
+            $this->request,
+            // FIXME Where do we get this list from? Is this something to be supplied by class implementor?
```

Comment:
I think it's a mistake to tunnel vision on the UIModelGrid use case here, remember, the API may also be used for automation. Is it not better to return everything and then decide in the HTML what to actually expose? Or perhaps the list of requested columns should be coming from the javascript as a paramter rather than hardcoded here.